repose-aggregator/services/service-client/impl/src/main/java/org/openrepose/core/services/serviceclient/akka/impl/AkkaServiceClientImpl.java

```diff
@@ -71,26 +74,49 @@ public AkkaServiceClientImpl(HttpClientService httpClientService) {
         quickFutureCache = CacheBuilder.newBuilder()
                 .expireAfterWrite(FUTURE_CACHE_TTL, FUTURE_CACHE_UNIT)
                 .build();
+    }
+
+    @PreDestroy
+    public void destroy() {
+        // tie this actor system to our bean lifecycle
+        actorSystem.shutdown();
+    }
 
+    @Override
+    public void setConnectionPoolId(String connectionPoolId) {
+        if (StringUtils.equals(this.connectionPoolId, connectionPoolId)) {
+            return;
+        }
 
-        tokenActorRef = actorSystem.actorOf(new Props(new UntypedActorFactory() {
-            public UntypedActor create() {
+        ActorRef oldTokenActorRef = tokenActorRef;
+        this.connectionPoolId = connectionPoolId;
+        createTokenActor(FORCE_CREATE);
 
-                return new AuthTokenFutureActor(serviceClient);
-            }
-        }).withRouter(new RoundRobinRouter(numberOfActors)), "authRequestRouter");
+        if (oldTokenActorRef != null) {
+            actorSystem.stop(oldTokenActorRef);
+        }
     }
 
+    private synchronized void createTokenActor(ForceCreate forceCreate) {
```

Comment:
Instead of trying to handle threads, you should just send messages to actors telling them to change their connection Pool ID. The actors will then change their own state, and no threading needs to happen. The actors can internally see if they are using the right connection pool, and if not change it. If they get multiple messages, it doesn't matter, the actor will know enough to throw it away.

Trying to manage it entirely outside of the actors seems like an error prone way to do this. Much better idea is to tell the actors that you want them to change state, and do it via a message. Then there aren't any synchronization issues, no attempting to hopefully create things in the proper thread.